### PR TITLE
feat: Add STACCube :gift:

### PR DESCRIPTION
**Pull request by fkroeber**
#### Description

This PR adds the STACCube class to the datacube.py. This is targeted at ad-hoc data cubes built from the results of a STAC metadata search. Contrary to the Opendatacube and the Geotiffarchive, it doesn't require pre-organising the data (e.g. to ingest the data into a database in case of the Opendatacube or to create a temporally stacked geotiff in case of the Geotiffarchive). Instead, the STACCube contains a retriever that knows how to fetch assets linked in STAC search results into a data cube. The further usage is demonstrated in the datacube.ipynb notebook.

#### Type of change
- [x] New feature (change which adds a feature requested in a specific issue)

#### Checklist
- [x] I have read and understood the [contributor guidelines](../CONTRIBUTING.md) of this project
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I ran all [demo notebooks](../demo) locally and there where no errors

### CODE REVIEW

**Comment by luukvdmeer** [comment]
This looks like a very useful addition, thanks! @loreabad6 can I add you as a reviewer?

**Comment by fkroeber** [comment]
The modifications refer to performance & functionality improvements realised via...
1. code refactoring: pulling data first from the STAC linked data src prior to daily resampling (group_by_solar_day)
2. introducing dtype parameters to allow data retrieval as integers instead of floats
3. adding dask_params to enable control over data retrieval (handled internally via stackstac library)

**Comment by loreabad6** [comment]
> This looks like a very useful addition, thanks! @loreabad6 can I add you as a reviewer?

Sure, I started checking and will continue tomorrow

File: semantique/datacube.py [hunk 1]
```diff
@@ -661,3 +663,320 @@ def _mask(self, data):
...
+
+        if self.config["group_by_solar_day"]:
+            if len(data.time):
+                days = data.time.astype("datetime64[D]")
+                if data.dtype.kind == "f":
+                    data = data.where(data != self.config["na_value"])
+                    data = data.groupby(days).first(skipna=True)
+                else:
+                    data = data.groupby(days).reduce(_mosaic_ints, na_value=self.config["na_value"])
+                data["time"] = data.time.astype("datetime64[D]").values
```
Comment (loreabad6) [inline semantique/datacube.py:932]: Handle nanosecond precision warning from pandas, there is maybe a better way for this
```suggestion
                days = data.time.astype("datetime64[ns]").dt.floor("D")
                if data.dtype.kind == "f":
                    data = data.where(data != self.config["na_value"])
                    data = data.groupby(days).first(skipna=True).rename({"floor": "time"})
                else:
                    data = data.groupby(days).reduce(_mosaic_ints, na_value=self.config["na_value"]).rename({"floor": "time"})
                data["time"] = data.time.values
```

File: semantique/datacube.py
Comment (loreabad6) [inline semantique/datacube.py:1]: Cosmetic change: indent with 2 spaces instead of 4

File: demo/files/layout_stac.json
Comment (loreabad6) [inline demo/files/layout_stac.json:1]: Cosmetic change: indent with 2 spaces instead of 4

**Comment by loreabad6** [comment]
Hi @fkroeber great PR! I made some mainly cosmetic suggestions. A couple of questions/comments coming from your comment above...

> 1. code refactoring: pulling data first from the STAC linked data src prior to daily resampling (group_by_solar_day)

I created a cloud-free composite with your implementation and if worked really nice, I just got the nanosecond precision for the handling of the groupping. I made a suggestion to handle it but maybe you know a better way

> 2. introducing dtype parameters to allow data retrieval as integers instead of floats

The integer handling is useful for using the SLC, but, from my tests, when working with the reflectance values we need to load as float, otherwise it does not work. I am a bit confused about that since the documented parameter says it will change it to float... in any case, does it then make sense to support int types?

> 3. adding dask_params to enable control over data retrieval (handled internally via stackstac library)

out of curiosity, xarray handles dask internally, right? have you experimented how semantique handles the chunk and data loading? is this just pass onto xarray or does semantique anyway request all the data, making the chunking in dask irrelevant?

**Review by loreabad6 — COMMENTED** [review]
Hi @fkroeber great PR! I made some mainly cosmetic suggestions. A couple of questions/comments coming from your comment above...

> 1. code refactoring: pulling data first from the STAC linked data src prior to daily resampling (group_by_solar_day)

I created a cloud-free composite with your implementation and if worked really nice, I just got the nanosecond precision for the handling of the groupping. I made a suggestion to handle it but maybe you know a better way

> 2. introducing dtype parameters to allow data retrieval as integers instead of floats

The integer handling is useful for using the SLC, but, from my tests, when working with the reflectance values we need to load as float, otherwise it does not work. I am a bit confused about that since the documented parameter says it will change it to float... in any case, does it then make sense to support int types?

> 3. adding dask_params to enable control over data retrieval (handled internally via stackstac library)

out of curiosity, xarray handles dask internally, right? have you experimented how semantique handles the chunk and data loading? is this just pass onto xarray or does semantique anyway request all the data, making the chunking in dask irrelevant?

File: demo/files/layout_stac.json [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+            "name": "coastal",
+            "description": "Coastal aerosol band (~443nm; 60m spatial resolution). Top of atmosphere (TOA) reflectance captured by the multi-spectral instrument (MSI) of Sentinel-2A or Sentinel-2B as band 1.",
+            "type": "continuous",
+            "values": {
+                "min": 0,
+                "max": 1,
+                "precision": 1
+            },
+            "copyright": "Contains modified Copernicus data."
+        },
```
[the remark this answers is not in the document]
Comment (fkroeber) [reply demo/files/layout_stac.json:13]: I'd like to refer you to the people who invented the layout file structure ;) The only thing I figured is that it is not necessary for any semantique functionality.

**Comment by fkroeber** [comment]
> Thanks @loreabad6 for taking time to review the changes! To refer to some comments/open questions...
  
> > 2. introducing dtype parameters to allow data retrieval as integers instead of floats
> 
> The integer handling is useful for using the SLC, but, from my tests, when working with the reflectance values we need to load as float, otherwise it does not work. I am a bit confused about that since the documented parameter says it will change it to float... in any case, does it then make sense to support int types?

Yes, when working with reflectance values, floats are necessary and not just for the recipe evaluation but already when fetching the data via STACCube. So in this case, the default paramter (float) should not be changed to int (maybe I should make this more clear in the documentation). The integer handling is really just to speed up the data fetching for layers that are natively in integer format. 

> > 3. adding dask_params to enable control over data retrieval (handled internally via stackstac library)
> 
> out of curiosity, xarray handles dask internally, right? have you experimented how semantique handles the chunk and data loading? is this just pass onto xarray or does semantique anyway request all the data, making the chunking in dask irrelevant?

Currently, semantique doesn't support dask parallelisation. The introduced dask_params for the STACCube really only refer on the data fetching part. Everything else (i.e. the main part of recipe execution) is not parallelised yet. 

As discussed, dask offers two main ways of parallelisation. Low-level xarray parallelisation (what you are refering to) and some high-level wrapper functionality (allowing to parallelise any function, similar to the working mechanism of map functions). Low-level parallelisation requires the xarray functions to support dask (currently not consistently ensured in semantique). I am currently working on a solution to have at least some workaround via high-level parallelisation to enhance the scalability of semantique. But more on this, in a different PR ;)

**Comment by fkroeber** [comment]
The minor fix regarding pixel alignment refers to the [following issue with stackstac](https://github.com/gjoseph92/stackstac/issues/168). With the modified parameters the retrieved data should now have exactly the same bounds as provided by the SpatialExtent object.

File: demo/files/layout_stac.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (whisperingpixel) [reply demo/files/layout_stac.json:13]: The **role** label in the layout was/is supposed to allow specifying meaning of bands. For example, which are the red/green/blue bands for the standard RGB true color visualisation in the frontend. It is not necessary for the functionality but could be used in cases where such additional labels are helpful.

**Pull request merged**